### ISSUE TITLE
Downloading Dojo/jQuery libs in Docker not required anymore

### ISSUE THREAD

**Issue by xezpeleta** [elkarbackup/elkarbackup#482]
As the upcoming version will include Dojo/jQuery in the repository, we don't need to download it in the Dockerfile build stage:

https://github.com/elkarbackup/elkarbackup/blob/dc7082c4896d47f6a95cb0573569d8bed1f35e88/docker/Dockerfile#L44-L61

**Comment by xezpeleta** [comment]
Same thing with the following `composer.json` scripts:


https://github.com/elkarbackup/elkarbackup/blob/dc7082c4896d47f6a95cb0573569d8bed1f35e88/composer.json#L39-L41

https://github.com/elkarbackup/elkarbackup/blob/dc7082c4896d47f6a95cb0573569d8bed1f35e88/composer.json#L45-L47


I understand we don't need this code if we include the js libs in the symfony public directory.

**Comment by elacunza** [comment]
I see that this was the case with dojo since almost the beginning, but not for jquery until 2.0.0 . Not very sure what should we do really; this impacts also bootstrap.sh from where we removed the download code for 2.0.0 .

Those downloads are not different from a compose install that we are performing during docker image/deb package builds... so maybe we should reintroduce them?

**Comment by elacunza** [comment]
> Same thing with the following `composer.json` scripts:
> 
> https://github.com/elkarbackup/elkarbackup/blob/dc7082c4896d47f6a95cb0573569d8bed1f35e88/composer.json#L39-L41
> 
> https://github.com/elkarbackup/elkarbackup/blob/dc7082c4896d47f6a95cb0573569d8bed1f35e88/composer.json#L45-L47
> 
> I understand we don't need this code if we include the js libs in the symfony public directory.

In this case (bootstrap) we're not uploading those files to the repository, we only have placeholder directories AFAIKS

**Comment by igorbga** [comment]
Disclaimer (sí, soy un chapas!) : If you don't feel like reading all my reflexion go to the last paragraph for my 2 cents.

That's somehitng that got us a little bit confused during the changes to 4.4, I think that as @elacunza already said this is something that comes from long ago and has never been resolved in a very consistent way.

To be honest I'm not sure right now which one would be the best approach (apart from completing the whole API and rewritting the whole frontend in React ;) ).

But I think there is a subtle but sustantial difference with vendor packages installed with composer and Jquery and Dojo. 

Composer is a dependency management solution mainly for the backend that uses a well stablished package repository (https://packagist.org/) so when you request it to "download" something you expect some degree of responsability in the side of the maintainers of packagist and the people that submit the packages there. You don't expect them to delete a concrete version arbitrarily just because there is already a better one, you also might expect them to follow some "semver" versioning strategy to allow you upgrade the version with more confidence and without breaking BC etc. There is a way of doing things and although we don't live in a happy pinky world most of time it works quite well.

In the other hand JQuery and Dojo are frontend libraries and we don't use a dependency management tool for frontend dependencies, we could, but we don't and honestly I doubt it would be worth the trouble with the actual frontend. 

There are solutions in order to include those frontend pieces through composer (https://stackoverflow.com/questions/11663559/how-to-install-jquery-with-composer, https://packagist.org/packages/maastermedia/dojo-bundle ) but they mostly are workarounds or little tricks to integrate everything in composer but by no way do they add the kind of "responsability" that I was talking about in previous paragraph, they just "do the download for you". 

Well in fact we are already using one of those "workarounds" (https://packagist.org/packages/twbs/bootstrap) to download Bootstrap CSS framework (don't confuse with our bootstraping script bootstrap.sh) but this one seems pretty solid as the version numbers seem to evolve in synchrony with Bootstrap's version and Bootstrap itself seems to use semantic versioning.

Ok, so much reflexion but let's get to the point, what is my proposal.

Those are IMHO the points to not include the dependencies in the GIT repo itself:

1- The size of the repo
2- Automatic updatability of non BC breaking versions
3- Licensing

But the first one really is not a big deal, they do take some space and requires a first download but not a big deal. And for the second one we are not really achieving it as we don't have a package manager, we always download the same version, don't we? The update of the version is completely on us. For the third one maybe we should check but given that JQuery, Dojo and Elkarbackup are all OpenSource projects I don't think that really is an issue.

Reasons to include them directly in the repo:

1- Do not risk for the maintainers to unpublish the version we are using
2- Ease build and deployment

I wouldn't give those a very high value, if the maintainer retire a version maybe we should check way and change the version not just use an old one and the build and deployment process is already "done". But that little value is more than nothing.

So to sum up, I would just donwload Dojo and JQuery, add them to the repo and be happy. Probably I would even do the same for Bootstrap CSS Framework in order to be consistent for all the dependency management for the frontend. Be aware that Bootstrap CSS Framework uses JQuery for some of it's features so probably there should be some kind of aligmente between Bootstrap and JQuery vesion.

**Comment by xezpeleta** [comment]
Probably the most purist solution is to keep them outside the repository. If we decide go that way, we may use a unified method to download these libraries for Debian packages and Docker images

The easiest solution would be to manage all this libraries using Composer too, but I think we don't have official repositories for all of them.

What about using yarn/npm to manage those external JS/CSS dependencies? It's probably the current standard way to do this job, but I'm not sure if it will help us in some way...

**Comment by elacunza** [comment]
I see both dojo and jquery version we use are in npm repos, so it seems a viable solution.

I think keeping downloads + SHA checks would keep it simpler and would require less work on our part (no idea about required work to use npm really)

Keeping them vendored directly in git as they're now, would be almost no work (except Docker build cleanup that would be nice).

I don't see a clear winner, but I don't think this provides any significant value because in all cases we'll ship them vendored in our binaries, so I'll agree to whatever decides the developer working on this... @xezpeleta ? :-)

**Comment by igorbga** [comment]
> What about using yarn/npm to manage those external JS/CSS dependencies? It's probably the current standard way to do this job, but I'm not sure if it will help us in some way...

I we end up going this route we should better use "yarn" as that is the package manager suggested for Webpack Encore, the Symfony component to integrate webpack and "modern" front-end parts in Symfony.

But as I said, right now i would bet on just adding them to our codebase.

**Comment by xezpeleta** [comment]
I've read your  ~research study~ post and finally you convinced me @igorbga

We stick to the simplest solution, including these files directly to our codebase.

PR: the best thing of rewritting the frontend with React would be to remove all the shitty js code I included years ago lol

**Comment by igorbga** [comment]
No freak, rest assured that React does also allow writting shitty code...
